core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/WorkflowCompiler.scala

```diff
@@ -36,6 +62,19 @@ class WorkflowCompiler(val workflowInfo: WorkflowInfo, val context: WorkflowCont
       .filter(pair => pair._2.nonEmpty)
 
   def amberWorkflow: Workflow = {
+    // pre-process: set output mode for sink based on the visualization operator before it
+    this.workflow.getSinkOperators.foreach(sinkOpId => {
+      val sinkOp = this.workflow.getOperator(sinkOpId)
+      val upstream = this.workflow.getUpstream(sinkOpId)
+      (upstream.head, sinkOp) match {
+        // match the combination of a visualization operator followed by a sink operator
+        case (viz: VisualizationOperator, sink: SimpleSinkOpDesc) =>
+          sink.setOutputMode(viz.outputMode())
+          sink.setChartType(viz.chartType())
+        case _ =>
+      }
+    })
+
```

Comment:
minor comment, as this `amberWorkflow` compilation method is getting bigger, maybe extract those chunks of codes as methods, for readability?

Reply:
Agree, we'll add more functionailities to workflow compiler later, such as checking the workflow integrity and give better error messages. As we add more in the future, we can do the refactor.